torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -1165,20 +1165,24 @@ void CudaCodeGen::call_raw(const std::vector<void*>& raw_args) {
       extent_args.push_back(raw_args[i]);
     }
   }
-  for (size_t i = 0; i < gpu_block_extents.size(); i++) {
-    if (gpu_block_extents[i]->isConstant()) {
-      gpu_block_extents_v[i] = immediateAs<int64_t>(gpu_block_extents[i]);
-      continue;
+  {
+    std::lock_guard<std::mutex> guard(eval_lock_);
+    for (size_t i = 0; i < gpu_block_extents.size(); i++) {
+      if (gpu_block_extents[i]->isConstant()) {
+        gpu_block_extents_v[i] = immediateAs<int64_t>(gpu_block_extents[i]);
+        continue;
+      }
+      gpu_block_extents_v[i] = block_extents_eval_[i].value<int64_t>(extent_args);
```

Comment:
Can you move the lock to just this statement? Most extents would be constants and they don't need to go through locking.